R/construction.R

```diff
@@ -1,66 +1,74 @@
 arrow_to_rdf = function(at, schema = NULL, schema_overrides = NULL, rechunk = TRUE) {
   # new column names by schema, #todo get names if schema not NULL
-  original_schema = schema
+  n_cols = at$num_columns
+
   new_schema = unpack_schema(
     schema = schema %||% at$ColumnNames(),
     schema_overrides = schema_overrides
   )
   col_names = names(new_schema)
 
-  if (length(col_names) != at$num_columns) stop("schema length does not match arrow table")
-
+  if (length(col_names) != n_cols) stop("schema length does not match arrow table")
 
-  # store special translated columns here
-  special_cols = new.env(parent = emptyenv())
+  data_cols = list()
+  # dictionaries cannot be built in different batches (categorical does not allow
+  # that) so we rechunk them and create them separately.
+  # struct columns don't work properly if they contain multiple chunks.
+  special_cols = list()
 
   ## iter over columns, possibly do special conversion
-  i_col = 0
-  for (column in at$columns) {
-    i_col = i_col + 1
-    name = col_names[i_col]
+  for (i in 1L:n_cols) {
+    column = at$column(i - 1L)
+    col_name = col_names[i]
+
     if (is_arrow_dictonary(column)) {
       column = coerce_arrow(column)
-      special_cols[[name]] = unwrap(arrow_to_rseries_result(name, column, rechunk))
-    } else if (is_arrow_struct(column) && has_multiple_chunks(column)) {
-      special_cols[[name]] = unwrap(arrow_to_rseries_result(name, column, rechunk))
-    } # else do nothing
+      special_cols[[col_name]] = unwrap(arrow_to_rseries_result(col_name, column, rechunk))
+    } else if (is_arrow_struct(column) && column$num_chunks > 1L) {
+      special_cols[[col_name]] = unwrap(arrow_to_rseries_result(col_name, column, rechunk))
+    } else {
+      data_cols[[col_name]] = column
+    }
   }
 
-  # drop already converted columns
-  at_new = at$SelectColumns(which(!names(at) %in% names(special_cols)) - 1L)
+  if (length(data_cols)) {
+    tbl = do.call(arrow::arrow_table, data_cols)
```

Comment:
don't we need to check if `arrow` is installed first since it's only a suggested dep?

Reply:
Good point, will be fixed.

Reply:
Thinking about it again, this function is only called from `as_polars_df.ArrowTabular`, so I don't think there is a possibility that it will be called without the arrow package.

Reply:
Indeed, thanks